### PR TITLE
reconcile mcg console pod

### PR DESCRIPTION
**Pull request by v-harihar**
This patch handles the reconcilation of mcg console pod.

Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

### CODE REVIEW

**Comment by v-harihar** [comment]
@naveenpaul1 Please review

**Comment by SanjalKatiyar** [comment]
Please change all the instances of `mcg-console` to `dfr-console` for now...

File: controllers/managedmcg_controller.go [hunk 1]
```diff
@@ -552,3 +568,38 @@ func (r *ManagedMCGReconciler) getCSVByPrefix(name string) (*opv1a1.ClusterServi
...
+	mcgConsolePlugin := console.GetConsolePluginCR(r.ConsolePort, basePath, OperatorNamespace)
+	_, err = controllerutil.CreateOrUpdate(context.TODO(), r.Client, mcgConsolePlugin, func() error {
+		return nil
+	})
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+
+	return nil
+}
```
Comment (SanjalKatiyar) [inline controllers/managedmcg_controller.go:605]: who is creating these resources ??
shouldn't this be done when operator is installed, we add deployment to operator CSV for that in case of ODF operator.

File: console/console.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+	consolev1alpha1 "github.com/openshift/api/console/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const MAIN_BASE_PATH = "/"
+const COMPATIBILITY_BASE_PATH = "/compatibility/"
```
Comment (SanjalKatiyar) [inline console/console.go:30]: no need for this COMPATIBILITY_BASE_PATH

File: controllers/managedmcg_controller.go (hunk 1 above)
Comment (SanjalKatiyar) [reply controllers/managedmcg_controller.go:605]: does here we are doing it via `config/console/kustomization.yaml` file ?

Comment (v-harihar) [reply controllers/managedmcg_controller.go:605]: @SanjalKatiyar you mean mcg console pod ? That is part of `bundle/manifests/mcg-osd-deployer.clusterserviceversion.yaml` as `deployments`

File: controllers/managedmcg_controller.go [hunk 3]
```diff
@@ -54,6 +57,7 @@ const (
 	deployerCSVPrefix = "mcg-osd-deployer"
 	noobaaFinalizer   = "noobaa.io/graceful_finalizer"
 	noobaaName        = "noobaa"
+	OperatorNamespace = "openshift-data-federation"
```
Comment (naveenpaul1) [inline controllers/managedmcg_controller.go:60]: try to get namespace from the request

File: controllers/managedmcg_controller.go [hunk 4]
```diff
@@ -194,6 +199,17 @@ func (r *ManagedMCGReconciler) reconcilePhases() (reconcile.Result, error) {
 		if err := r.reconcileOCSCSV(); err != nil {
 			return ctrl.Result{}, err
 		}
+
+		logger := log.FromContext(ctx)
+		instance := configv1.ClusterVersion{}
+		if err := r.Client.Get(context.TODO(), req.NamespacedName, &instance); err != nil {
+			return ctrl.Result{}, err
+		}
```
Comment (SanjalKatiyar) [inline controllers/managedmcg_controller.go:207]: no need right now...

File: controllers/managedmcg_controller.go (hunk 3 above)
Comment (v-harihar) [reply controllers/managedmcg_controller.go]: Hard coding it to `redhat-data-federation` as of now.

Comment (naveenpaul1) [reply controllers/managedmcg_controller.go]: @v-harihar why we need to hardcode the namespace, can we get it from `ManagedMCGReconciler.namespace`

File: controllers/managedmcg_controller.go (hunk 4 above)
Comment (v-harihar) [reply controllers/managedmcg_controller.go:207]: The cluster version that is fetched here used to get the base_path.
About `base_path` usability:
```
basePath is the path to the plugin's assets. The primary asset it the
manifest file called `plugin-manifest.json`, which is a JSON document
that contains metadata about the plugin and the extensions.
```
Let's keep the infra as is and option open.

File: console/console.go (hunk 2 above)
Comment (v-harihar) [reply console/console.go:30]: Answered this in https://github.com/red-hat-storage/mcg-osd-deployer/pull/16#discussion_r865653546

File: console/console.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+				Namespace: serviceNamespace,
+				Port:      int32(consolePort),
+				BasePath:  basePath,
+			},
+		},
+	}
+}
+
+func GetBasePath(clusterVersion string) string {
+	if strings.Contains(clusterVersion, "1.0.0") {
```
Comment (SanjalKatiyar) [inline console/console.go:88]: if we need to keep this "1.0.0" not a correct version... this is OCP version... 
```suggestion
	if strings.Contains(clusterVersion, "4.11") {
```

Comment (rexagod) [inline console/console.go:88]:
```suggestion
	if strings.Contains(clusterVersion, "4.11") {
```

**Review by rexagod — CHANGES_REQUESTED** [review]
~~Will review once more after the PR is good to go.~~ (Suggested the same change as above, @v-harihar PTAL).

File: console/console.go (hunk 5 above)
Comment (v-harihar) [reply console/console.go:88]: done!

File: config/console/console_init.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+              readOnly: true
+      tolerations:
+      - effect: NoSchedule
+        key: node.ocs.openshift.io/storage
+        operator: Equal
+        value: "true"
+      volumes:
+        - name: addon-mcg-osd-parameters
+          secret:
+            secretName: addon-mcg-osd-parameters
```
Comment (v-harihar) [inline config/console/console_init.yaml:36]: This is a temp secretName, just to get the things up and running

File: controllers/managedmcg_controller.go (hunk 3 above)
Comment (rexagod) [reply controllers/managedmcg_controller.go]: +1

File: config/console/console_init.yaml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    matchLabels:
+      app: mcg-ms-console
+  template:
+    metadata:
+      labels:
+        app: mcg-ms-console
+    spec:
+      containers:
+        - name: mcg-ms-console
+          image: mcg-ms-console:latest
```
Comment (naveenpaul1) [inline config/console/console_init.yaml:16]: can we give the image like this? it will check in local machine and dockerHub.

File: config/console/console_init.yaml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+          resources:
+            limits:
```
Comment (naveenpaul1) [inline config/console/console_init.yaml:18]: can we add "request" also here?

File: console/console.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+				Namespace: serviceNamespace,
+				Port:      int32(consolePort),
+				BasePath:  basePath,
+			},
+		},
+	}
+}
+
+func GetBasePath(clusterVersion string) string {
+	if strings.Contains(clusterVersion, "4.11") {
```
Comment (naveenpaul1) [inline console/console.go:88]: only for version 4.11 we need to give a specific path, is that right?

File: config/console/console_init.yaml (hunk 7 above)
Comment (v-harihar) [reply config/console/console_init.yaml:16]: What I see is, the path referred to fetch the image is built in `kustomization.yaml`. https://github.com/red-hat-storage/mcg-osd-deployer/blob/f465a00894ee87655f57f26309d7f0e29e6f53c2/config/console/kustomization.yaml#L13-L15

File: config/console/console_init.yaml (hunk 8 above)
Comment (v-harihar) [reply config/console/console_init.yaml:18]: In odf/console, we dont have requests. Does it mean it runs with lower resources as well but limit is required to have boundary condition ?

File: console/console.go (hunk 9 above)
Comment (SanjalKatiyar) [reply console/console.go:88]: yeah... but it is not required exactly right now... may be in future when we use OCP 4.11... then we need to make changes to mcg-ms-console as well...
for first release we are focusing on OCP 4.10... so we can either keep it or remove it whatever is alright...

File: config/console/kustomization.yaml [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+  app: mcg-ms-console
+
+resources:
+- console_init.yaml
+
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+images:
+- name: mcg-ms-console
+  newName: quay.io/skatiyar/mcg-ms-console
```
[the remark this answers is not in the document]
Comment (SanjalKatiyar) [reply config/console/kustomization.yaml:14]: plz let me know and I will push the image to that repo (if there is any common place for development purpose)...
also once build team comes into the picture I think we don't we will have such issues !!

File: config/console/console_init.yaml (hunk 7 above)
Comment (naveenpaul1) [reply config/console/console_init.yaml:16]: @v-harihar yes correct

File: console/console.go (hunk 9 above)
Comment (rexagod) [reply console/console.go:88]: I think this is something we need to take a step back and think about. Do we include these changes right now (given that we do not need them at this very moment, but might in the future), or do we remove them and just merge the bare minimum that's required to reconcile UI plugin with the operator?

Comment (v-harihar) [reply console/console.go:88]: > only for version 4.11 we need to give a specific path, is that right?

yes.

Comment (v-harihar) [reply console/console.go:88]: > I think this is something we need to take a step back and think about. Do we include these changes right now (given that we do not need them at this very moment, but might in the future), or do we remove them and just merge the bare minimum that's required to reconcile UI plugin with the operator?


Just want to quote the basepath definition here:

```
// basePath is the path to the plugin's assets. The primary asset it the
// manifest file called `plugin-manifest.json`, which is a JSON document
// that contains metadata about the plugin and the extensions.
```

In versions < 4.11, the metadata file is stored under `/`, later to 4.11 specific path can be provided. It doesn't have any special reason to not be there as it points to a place holder and it is just a part of infra provided by the CR. So it is up to the pod whether it uses it or ignores it.

Comment (SanjalKatiyar) [reply console/console.go:88]: `In versions < 4.11, the metadata file is stored under /, later to 4.11 specific path can be provided.` --> **this is wrong**
in 4.11 as well basepath can be `/`, basically it depends...
in ODF we have 3 branches example: `release-4.10` and `release-4.10-compatibility` and `release-4.11`
OCP 4.11 when uses ODF 4.11 --> basepath is `/` 
OCP 4.11 when uses ODF 4.10 --> basepath is `/compatibility`
OCP 4.10 when uses ODF 4.10 --> basepath is again `/`

`It doesn't have any special reason to not be there as it points to a place holder and it is just a part of infra provided by the CR. So it is up to the pod whether it uses it or ignores it.` --> I agree... it is up to you guys whether you want to add it or not... anything is fine... these are just few lines.... when we support mcg-ms-console with OCP 4.11 we will have to revisit it later...

File: config/console/kustomization.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (SanjalKatiyar) [reply config/console/kustomization.yaml:14]: I don't think I have access, but I think we should not block this PR on this. We can always change it later.
Or if someone has access they can push the image themselves, we just need:
https://github.com/red-hat-storage/mcg-ms-console/blob/master/Dockerfile
https://github.com/red-hat-storage/mcg-ms-console/blob/master/default.conf 
That is all, it will do the cloning, building and everything. And we can push that to any repository.

File: console/console.go (hunk 9 above)
Comment (rexagod) [reply console/console.go:88]: ACK, sounds good.

File: config/console/kustomization.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (naveenpaul1) [reply config/console/kustomization.yaml:14]: yes @SanjalKatiyar it's not a blocker, just a suggestion

File: controllers/managedmcg_controller.go [hunk 11]
```diff
@@ -594,3 +610,38 @@ func (r *ManagedMCGReconciler) getCSVByPrefix(name string) (*opv1a1.ClusterServi
...
 	return csv, nil
 }
+
+func (r *ManagedMCGReconciler) ensureConsolePlugin(clusterVersion string) error {
+	// The base path to where the request are sent
+	basePath := console.GetBasePath(clusterVersion)
+
+	// Get mcg console Deployment
+	mcgConsoleDeployment := console.GetDeployment(r.namespace)
+	err := r.Client.Get(context.TODO(), types.NamespacedName{
```
Comment (rexagod) [inline controllers/managedmcg_controller.go:620]:
```suggestion
	err := r.Client.Get(r.ctx, types.NamespacedName{
```
Please update all such instances.

File: controllers/managedmcg_controller.go [hunk 12]
```diff
@@ -594,3 +610,38 @@ func (r *ManagedMCGReconciler) getCSVByPrefix(name string) (*opv1a1.ClusterServi
...
+	if err != nil {
+		return err
+	}
+
+	// Create/Update mcg console Service
+	mcgConsoleService := console.GetService(r.ConsolePort, r.namespace)
+	_, err = controllerutil.CreateOrUpdate(context.TODO(), r.Client, mcgConsoleService, func() error {
+		return controllerutil.SetControllerReference(mcgConsoleDeployment, mcgConsoleService, r.Scheme)
+	})
+	if err != nil && !errors.IsAlreadyExists(err) {
```
Comment (rexagod) [inline controllers/managedmcg_controller.go:633]:
```suggestion
	if err != nil {
```
`CreateOrUpdate` will not error out if the object exists, it'll just proceed to update it in that case.

File: controllers/managedmcg_controller.go [hunk 13]
```diff
@@ -594,3 +610,38 @@ func (r *ManagedMCGReconciler) getCSVByPrefix(name string) (*opv1a1.ClusterServi
...
+		return err
+	}
+
+	// Create/Update mcg console ConsolePlugin
+	mcgConsolePlugin := console.GetConsolePluginCR(r.ConsolePort, basePath, r.namespace)
+	_, err = controllerutil.CreateOrUpdate(context.TODO(), r.Client, mcgConsolePlugin, func() error {
+		return nil
+	})
+	if err != nil && !errors.IsAlreadyExists(err) {
```
Comment (rexagod) [inline controllers/managedmcg_controller.go:642]:
```suggestion
	if err != nil {
```
`CreateOrUpdate` will not error out if the object exists, it'll just proceed to update it in that case.

File: hack/make-bundle-vars.mk [hunk 14]
```diff
@@ -0,0 +1,6 @@
+
+MCG_CONSOLE_IMG_NAME ?= mcg-ms-console
+MCG_CONSOLE_IMG_TAG ?= latest
+MCG_CONSOLE_IMG_LOCATION ?= quay.io/skatiyar
+MCG_CONSOLE_IMG ?= $(MCG_CONSOLE_IMG_LOCATION)/$(MCG_CONSOLE_IMG_NAME):$(MCG_CONSOLE_IMG_TAG)
```
Comment (rexagod) [inline hack/make-bundle-vars.mk:5]: Can we move them to `.env` (since these are env vars)? Any reason not to?

**Review by rexagod — CHANGES_REQUESTED** [review]

**Comment by rexagod** [comment]
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/red-hat-storage_mcg-osd-deployer/16/pull-ci-red-hat-storage-mcg-osd-deployer-main-lint/1526836416699437056#1:build-log.txt%3A55-60

File: console/console.go [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+	}
+}
+
+func GetConsolePluginCR(consolePort int, basePath string, serviceNamespace string) *consolev1alpha1.ConsolePlugin {
+	return &consolev1alpha1.ConsolePlugin{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "mcg-ms-console",
+		},
+		Spec: consolev1alpha1.ConsolePluginSpec{
+			DisplayName: "mcg Plugin",
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply console/console.go:76]: May be this is better `MCG-managed-service`

File: console/console.go [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+	}
+}
+
+func GetService(port int, namespace string) *apiv1.Service {
+	return &apiv1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "mcg-ms-console-service",
+			Namespace: namespace,
+			Annotations: map[string]string{
+				//TODO: Replace "addon-mcg-osd-parameters" with appr. secret
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply console/console.go:47]: > mcg-ms-console-service?

I think it is all right

File: controllers/managedmcg_controller.go [hunk 17]
```diff
@@ -121,7 +126,7 @@ func (r *ManagedMCGReconciler) initializeReconciler(req ctrl.Request) {
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-func (r *ManagedMCGReconciler) Reconcile(_ context.Context, req ctrl.Request) (ctrl.Result, error) {
+func (r *ManagedMCGReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply controllers/managedmcg_controller.go:129]: We may also need to clean up 1st received arg, that has no sense to keep it there. In this patch I am just falling back to existing args pass and use ctx from `r`

File: hack/make-bundle-vars.mk (hunk 14 above)
Comment (v-harihar) [reply hack/make-bundle-vars.mk:5]: These are not env vars but the tag names used in main Make file. This way of defining tags in seperate file will leave main Makefile clean

File: controllers/managedmcg_controller.go [hunk 18]
```diff
@@ -152,7 +157,7 @@ func (r *ManagedMCGReconciler) Reconcile(_ context.Context, req ctrl.Request) (c
 	}
 }
 
-func (r *ManagedMCGReconciler) reconcilePhases() (reconcile.Result, error) {
+func (r *ManagedMCGReconciler) reconcilePhases(ctx context.Context, req ctrl.Request) (reconcile.Result, error) {
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply controllers/managedmcg_controller.go:160]: `req ctrl.Request` is needed to fetch the `namespace` from the `req` received by `Reconcile`. This avoids hard coding of name space

File: controllers/managedmcg_controller.go [hunk 19]
```diff
@@ -194,6 +199,17 @@ func (r *ManagedMCGReconciler) reconcilePhases() (reconcile.Result, error) {
...
 		}
+
+		logger := log.FromContext(ctx)
+		instance := configv1.ClusterVersion{}
+		if err := r.Client.Get(context.TODO(), req.NamespacedName, &instance); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if err := r.ensureConsolePlugin(instance.Status.Desired.Version); err != nil {
+			logger.Error(err, "Could not ensure compatibility for MCG consolePlugin")
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply controllers/managedmcg_controller.go:210]: sure

File: controllers/managedmcg_controller.go (hunk 11 above)
Comment (v-harihar) [reply controllers/managedmcg_controller.go:620]: IMO, it is always better to fetch the ctx from methods that saved one. I will change it suggested way to keep this things matched

File: controllers/managedmcg_controller.go [hunk 20]
```diff
@@ -194,6 +199,17 @@ func (r *ManagedMCGReconciler) reconcilePhases() (reconcile.Result, error) {
 		if err := r.reconcileOCSCSV(); err != nil {
 			return ctrl.Result{}, err
 		}
+
+		logger := log.FromContext(ctx)
+		instance := configv1.ClusterVersion{}
+		if err := r.Client.Get(context.TODO(), req.NamespacedName, &instance); err != nil {
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply controllers/managedmcg_controller.go:205]: IMO, it is always better to fetch the ctx from methods than the saved one. Because the initialized/saved context may become invalid . Also in our code we are not using the ctx to control the flow. 

I will change it suggested way to keep this things matched

File: controllers/managedmcg_controller.go [hunk 21]
```diff
@@ -594,3 +610,38 @@ func (r *ManagedMCGReconciler) getCSVByPrefix(name string) (*opv1a1.ClusterServi
 
 	return csv, nil
 }
+
+func (r *ManagedMCGReconciler) ensureConsolePlugin(clusterVersion string) error {
+	// The base path to where the request are sent
```
[the remark this answers is not in the document]
Comment (v-harihar) [reply controllers/managedmcg_controller.go:615]: It is already explained in detail as part of `ConsolePluginService`.

File: hack/make-bundle-vars.mk (hunk 14 above)
Comment (rexagod) [reply hack/make-bundle-vars.mk:5]: I misspoke, not env vars, just vars. But what I meant was, these are defined and used in the same way as https://github.com/red-hat-storage/mcg-osd-deployer/blob/main/shim/.env, are they not?

File: controllers/managedmcg_controller.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (rexagod) [reply controllers/managedmcg_controller.go:613]: Can you point me there? I still anyone reading this should have some indication where it is, or better yet the comment should be elaborated to say so.

File: hack/make-bundle-vars.mk (hunk 14 above)
Comment (v-harihar) [reply hack/make-bundle-vars.mk:5]: Its more of format. Even if you check https://github.com/red-hat-storage/odf-operator/blob/main/hack/make-bundle-vars.mk, where Makefile vars/tags are part of the similar file. `.mk` is one such file format used to have such parameters.

File: controllers/managedmcg_controller.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (v-harihar) [reply controllers/managedmcg_controller.go:613]: You dont need to elaborate it here. The variable what we are referring is explained inside the `ConsolePluginService` struct. Here it is https://github.com/openshift/api/blob/ea6292c6ebaad8cacd0e6d2ef45d91585f66738c/console/v1alpha1/types_console_plugin.go#L129

File: config/console/console_init.yaml (hunk 6 above)
Comment (rexagod) [inline config/console/console_init.yaml:36]: I think the reason why the TLS artifacts are not being created is because we didn't create a `secret` with this name anywhere in the first place. If you look [here](https://dev.to/joshduffney/kubernetes-using-configmap-subpaths-to-mount-files-3a1i), they use a similar approach (`configmap` instead of `secret`, but it does not matter since the created files will have the name as the `configmap/secret.Data`'s keys and the contents as the respective values for those keys).

Basically, we need to make sure that the dummy `secret` is present with the required data before the execution reaches this point.

Comment (v-harihar) [reply config/console/console_init.yaml:36]: Yes, we need to create the secret externally when we run it manually. sample secret:

```
kind: Secret
apiVersion: v1
metadata:
  name: addon-mcg-osd-parameters
  namespace: fed
data:
  tls.crt: "base64 encoded cert"
  tls.key: "base64 encoded key"
type: kubernetes.io/tls
```

**Review by naveenpaul1 — APPROVED** [review]
/LGTM
